Add basic functions for colorspace transforms.

SomTambe opened the pull request:
This PR makes the base for any colorspace transform.
- add `colorify` and `channelify` functions.

Usage: 
```julia
julia> f = Chain(x->HSV.(x),channelify,flatten,Dense(768,16),Dense(16,10),x->σ.(x))

julia> f(rand(RGB,16,16,1)) # single image of dims 16 x 16
```
`colorify` can be used **if** one ever requires a colorspace transform between two layers.
```julia
julia> f = Chain(...,
                         x->colorify(RGB,x),
                         x->Gray.(x),
                         channelify,
                         ...
                         ) # this would allow for a Grayscale colorspace transform in between the network
```
I am not sure if ever this will come to use. It's better to provide an option rather than not providing anything at all.
I would love to hear your comments @DhairyaLGandhi @johnnychen94 

Also, `channelview` is not differentiable. I will be writing an adjoint for that.

johnnychen94 commented on src/colors/conversions.jl:
```diff
@@ -0,0 +1,43 @@
+"""
+function channelify(m::AbstractArray{T}) where T <: Color
+    m = channelview(m)
+    t = collect(1:ndims(m))
```
use `Tuple` whenever possible

```suggestion
    t = ntuple(_->i, ndims(m))
```

johnnychen94 reviewed (commented):
This is a good start.

One thing I notice about this PR is that `permutedims` creates a memory contiguous array while `colorview` doesn't. This implies that `channeify` creates a contiguous dense array, while `colorify` does not.

Perhaps a 1-argument version of `colorify` would be useful, i.e., `colorify(RGB)` that creates `x->colorify(RGB, x)`.

Should this function specially handle `ndims(m) == 2`? I guess no? If so it's probably worth documenting something like "the last dimension is faithfully recognized as a batch dimension so you need to unsqueeze the input manually if the input is only one sample"

SomTambe commented:
>Should this function specially handle ndims(m) == 2? I guess no? If so it's probably worth documenting something like "the last dimension is faithfully recognized as a batch dimension so you need to unsqueeze the input manually if the input is only one sample"

Yes you are right. I do require them to have the last dimension as a batch. I should document it as you say.

johnnychen94 commented on src/colors/conversions.jl:
```diff
@@ -0,0 +1,45 @@
+function channelify(m::AbstractArray{T}) where T <: Color
+    m = channelview(m)
+    t = ntuple(i->i, ndims(m))
+    if T <: AbstractGray
+        m = unsqueeze(m, t[end-1])
+    end
```
For the record, this seemly inconsistent behavior for `Gray` is decided in https://github.com/JuliaImages/ImageCore.jl/issues/97 so it's not going to change.

SomTambe replied:
This has been specially done because Flux required the dimensions to be in `(height, width, channels, batch)` format, so it would be inconsistent with Flux.

SomTambe commented:
@johnnychen94 Should I merge this PR? I'll add the adjoints in another PR, will that be fine?

johnnychen94 commented:
Every non-trivial PR should have its associated test cases.

SomTambe commented:
>Every non-trivial PR should have its associated test cases.

I completely forgot about the tests. No worries I will write them.

SomTambe commented:
@johnnychen94 I have written the tests. Could you review them please?

johnnychen94 reviewed (commented):
I feel this unit test is unnecessarily complicated. You don't need to introduce all other layers as long as the basic properties of `channelify`/`colorify` get tested.

Why not just test

```julia
X = rand(RGB{N0f8}, 4, 4, 5)
Y = channelify(X)

@test Y[:, :, 1, 1] == channelview(X[:, :, 1])
@test size(Y) == (4, 4, 3, 5)
@test eltype(Y) == N0f8
# make sure it has contiguous memory layout
@test stride(Y, 1) == 1
# maybe there're other properties
```

and also test 2d, 3d and 4d cases. And test various colorspace, e.g., `RGB`, `HSV`, `Gray`, `RGBA`, `Gray24`.

SomTambe commented:
I have added the adjoints for the necessary functions. I'll be adding the tests by tonight too.
cc. @DhairyaLGandhi @johnnychen94

DhairyaLGandhi commented on src/colors/conversions.jl:
```diff
@@ -0,0 +1,89 @@
+    if e <: AbstractGray
+        m = unsqueeze(m, 1)
+    end
+    t = ntuple(i->i, ndims(m))
```
```suggestion
    t = ntuple(identity, ndims(m))
    # or
    t = 1:N
```

DhairyaLGandhi commented on src/colors/conversions.jl:
```diff
@@ -0,0 +1,89 @@
+
+"""
+function channelify(m::AbstractArray{T}) where T <: Color
+    e = eltype(m)
```
This is just `T`

DhairyaLGandhi replied:
Either way, you don't need to call eltype explicitly here and in other places

DhairyaLGandhi commented on src/colors/conversions.jl:
```diff
@@ -0,0 +1,89 @@
+Output will be in the order ``(*, channels, batch)``.
+
+"""
+function channelify(m::AbstractArray{T}) where T <: Color
```
```suggestion
function channelify(m::AbstractArray{T,N}) where {T <: Color, N}
```

DhairyaLGandhi commented on src/colors/conversions.jl:
```diff
@@ -0,0 +1,89 @@
+end
+
+#adjoint for channelview
+Zygote.@adjoint channelview(x) = begin
```
```suggestion
@adjoint function channelview(x::AbstractArray{T,N}) where {T,N}
```

DhairyaLGandhi replied:
Also import `@adjount` since you're likely to need that and `nograd` frequently

DhairyaLGandhi commented on src/colors/conversions.jl:
```diff
@@ -0,0 +1,89 @@
+end
+
+# adjoint for colorview
+Zygote.@adjoint colorview(T,x) = begin
```
See above

DhairyaLGandhi commented on src/colors/conversions.jl:
```diff
@@ -0,0 +1,89 @@
+          :YIQ,:AYIQ,:YIQA)
+    @eval Zygote.@adjoint $f(x...) = begin
+        y = $f(x...)
+        function pullback(Δ)
```
```suggestion
        function $f_pullback(Δ)
```
?

SomTambe replied:
`$f` will just append the name of the function `f` to the pullback function right? So what difference will it make?

johnnychen94 replied:
That looks like a convention to me. Also `Zygote.pullback` is a function.

SomTambe replied:
For some reason, Zygote isn't being able to parse this. I'll just change the name of the function to `pull`.

DhairyaLGandhi commented on tests/colors/conversions.jl:
```diff
@@ -0,0 +1,30 @@
+    @testset "Testing colorify" begin
+        @test size(g_3(rand(20,20,7,3))) == (10,3)
+        @test size(g_1(rand(20,20,7,3))) == (10,3)
+    end
```
Also test for differentiability

SomTambe replied:
Yes sure

johnnychen94 commented on src/colors/conversions.jl:
```diff
@@ -0,0 +1,90 @@
+    @eval @adjoint function $f(x...)
+        y = $f(x...)
+        function pullback(Δ)
+            return (Δ,)
```
This pullback doesn't look right to me. Maybe just `Δ` (since we need to expand it into `x...`)?

SomTambe replied:
Yup corrected it.

SomTambe commented on src/colors/conversions.jl:
```diff
@@ -0,0 +1,90 @@
+          :DIN99,:ADIN99,:DIN99A,
+          :LMS,:ALMS,:LMSA,
+          :YIQ,:AYIQ,:YIQA)
+    @eval @adjoint function $f(x...)
```
Why should it be for Real only?

johnnychen94 replied:
My first look is that the pullback is not correct for the `Colorant` inputs.

The most important thing currently is to _add unit tests_. More than 70% of the efforts should be paid to unit tests to make sure things are written correctly.

With unit tests appropriately written, we can then revisit this one and see if it is correct for colorant inputs.

johnnychen94 commented on tests/colors/conversions.jl:
```diff
@@ -1,30 +1,67 @@
+                       :AGray,:GrayA,:Gray, 
+                       :HSV) # colorspaces those have samplers defined for Base.Random
 
+    cspaces = (:HSV,:AHSV,:HSVA,
+               :Gray,:AGray,:GrayA,
+               :HSL,:AHSL,:HSLA,
+               :RGB,:ARGB,:RGBA,
+               :BGR,:ABGR,:BGRA,
+               :XYZ,:AXYZ,:XYZA,
+               :xyY,:AxyY,:xyYA,
+               :Lab,:ALab,:LabA,
+               :Luv,:ALuv,:LuvA,
+               :LCHab,:ALCHab,:LCHabA,
+               :LCHuv,:ALCHuv,:LCHuvA,
+               :DIN99,:ADIN99,:DIN99A,
+               :LMS,:ALMS,:LMSA,
+               :YIQ,:AYIQ,:YIQA) # tuple of all exprs of Colorspaces and their transparent variants
```
You don't need to write them as `Symbol` and `rand(eval(cs), ds...)`, you can simply use plain types and use `rand(cs, ds...)`.

SomTambe replied:
Absolutely, I am just in the process of writing the `colorify` tests and then the differentiablilty tests. Goog thing you reminded me about reversibility, I had totally forgotten about it :)

johnnychen94 reviewed (commented):
A good start on improving coverage.

DhairyaLGandhi commented on src/colors/conversions.jl:
```diff
@@ -0,0 +1,90 @@
+end
+
+# adjoint for colorview
+@adjoint function colorview(T, x) where {T}
```
Does colorview handle batches?

SomTambe replied:
Yes it does

johnnychen94 replied:
Good question. Yes it does; `channelview` and `colorview` is just a wrapper of `Base.ReinterpretedArray`:

```
julia> X = rand(RGB, 12, 24, 64);

julia> channelview(X) |> size
(3, 12, 24, 64)

julia> colorview(eltype(X), channelview(X)) == X
true
```

SomTambe commented:
Should we get this merged for the while? Differentiability is still an issue for `(CT<:Colorant)(x::T) where T` and it's taking a lot of debugging for the adjoints to be written for these, which I feel would be not required since colorspace transforms are not at all required for the Geometry stuff we are at.
I will accordingly open an issue for the `adjoints` of `(CT<:Colorant)(x::T) where T` and we can do that a while later.
CC: @johnnychen94 @DhairyaLGandhi

johnnychen94 commented:
It's fine to separate the adjoint to another future PR. But I don't think this PR is finished.

As a summary, here's what I think should be covered in this PR:

- [ ] differentiability test for `colorview` and `channelview`. Or equivalently, test `colorify` and `channelify`. https://github.com/SomTambe/DiffImages.jl/pull/3#discussion_r650553436
- [ ] There is a multi-input version of `colorview` that needs an adjoint implementation: `colorview(T, gray1, gray2...)` https://github.com/SomTambe/DiffImages.jl/pull/3#discussion_r650557738 If you don't plan to support it in this PR, leave a comment in the source code.
- [ ] If the adjoint of `(CT<:Colorant)(x::T) where T` is still an issue. They should be removed for this PR and added back in future PR.
- [ ] set up CI and make sure it's all green checks.
- [ ] `jldoctest` in the docstring seems broken, make sure it passes. You can add jldoctest as a part of the unit test by calling `Documenter.doctest(DiffImages, manual = false)`

You might also need to bookmark what's unsolved so we don't need to repeat the discussion in the new PR:

- [ ] `(CT<:Colorant)(x::T) where T<:Colorant` for is missing https://github.com/SomTambe/DiffImages.jl/pull/3#discussion_r650557355
- [ ] if https://github.com/SomTambe/DiffImages.jl/pull/3#discussion_r650557922 solves the issue in https://github.com/FluxML/Zygote.jl/issues/993

I might miss some unsolved issues. When the PR growing too big, it's always a good practice to provide a summary of the current status.

SomTambe commented:
Sounds good. I had totally forgotten for writing tests for `colorview` and `channelview`. I'll do that and set up CI + doctests for the PR.

SomTambe commented:
>If the adjoint of (CT<:Colorant)(x::T) where T is still an issue. The relevant codes should be removed for this PR and added back in future PR.

I have commented out the tests which would evaluate gradients for these. This would be added in the next PR.
- [x] Added gradients for `colorview` and `channelview`.
- [ ] Completed CI integration.

I am still debugging CI integration. Will be done soon.

SomTambe commented:
@johnnychen94 I have done all the required tasks we needed. I have also created #5 to track all our progress remaining over here.

johnnychen94 reviewed (approved):
Last round of comments with only style suggestions.

@DhairyaLGandhi any more comments?

SomTambe commented:
Committed all the changes you asked for.

johnnychen94 commented on .github/workflows/ci.yml:
```diff
@@ -42,8 +42,15 @@ jobs:
       - run: |
           julia --project=docs -e '
             using Pkg
+            Pkg.add("Documenter")
```
This one can also be added into `docs/Project.toml`

johnnychen94 commented on .github/workflows/ci.yml:
```diff
@@ -42,8 +42,15 @@ jobs:
+            Pkg.add("Documenter")
             Pkg.develop(PackageSpec(path=pwd()))
             Pkg.instantiate()'
+      - run: |
+          julia --project=docs -e '
+            using Pkg
+            using Documenter: doctest
+            using DiffImages
+            doctest(DiffImages)'
```
Sorry that I didn't make it clear. I was thinking of adding `doctest(DiffImages)` into `test/runtests.jl`. See [OffsetArrays](https://github.com/JuliaArrays/OffsetArrays.jl/blob/7a43b4ca2b10a8c24a162f1365976d6a227a1666/test/runtests.jl#L42-L48) as an example.

DhairyaLGandhi commented on src/colors/conversions.jl:
```diff
@@ -0,0 +1,105 @@
+function channelify(m::AbstractArray{CT,N}) where {CT <: Colorant, N}
+    e = eltype(m)
+    m = channelview(m)
+    if e <: AbstractGray
```
Can this not be handled by a generic reshape? Those would be no-ops in the remaining cases

SomTambe replied:
Yes it is a simple reshape. I'll change it to a reshape then.

DhairyaLGandhi commented on test/colors/conversions.jl:
```diff
@@ -0,0 +1,155 @@
+                    @test gradient(x->sum(channelview(colorview(cs,x))),i)[1]==ones(size(i))
+                else
+                    i = rand(ds3[end - 1], ds3[1:end - 2]..., ds3[end])
+                    @test gradient(x->sum(channelview(colorview(cs,x))),i)[1]==ones(size(i))
```
A cleaner design might be to use the if else tree to get `i` or the args and testing gradient only once.


Also you should add tests for differentiability of colorify, channelify and have separate tests for channelview etc as well

SomTambe replied:
I already had tests for `channelview` and `colorview`. Do I need to write the `channelify` and `colorify` differentiability tests?

DhairyaLGandhi replied:
I just mean that testing them as individual kernels and verifying they do the right thing is good practice in unit testing.

DhairyaLGandhi reviewed (commented):
Largely LGTM , added a couple thoughts on the tests and let's get this merged!

SomTambe commented:
I am merging the PR as per conversation with @DhairyaLGandhi 
cc @johnnychen94

DhairyaLGandhi commented on src/colors/conversions.jl:
````diff
@@ -0,0 +1,105 @@
+(25, 25, 7)
+```
+"""
+function colorify(color::Type{CT}, m::AbstractArray) where CT <: Colorant
````
I feel like we may want to restrict this function to `AbstractArray{T,4}` since the case of `(28,28,3)` with Grayscale should be a simple reinterpret

SomTambe replied:
The current function gives the freedom of any number of dims >= 2. We could restrict it to 4 but it could be we may require something for higher dimensions in the future.

DhairyaLGandhi replied:
My main concern is the case I highlighted earlier. If that is handled correctly, then I'm fine

johnnychen94 replied:
I understand that classic deep learning mainly deals with 4D tensor but I'm not very positive about this type constraint; especially when it's not due to implementation restriction.

DhairyaLGandhi replied:
I am the first person to not restrict types for sure, just wanted to ensure we're doing the correct thing for different inputs we can expect to get.
